core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java

```diff
@@ -0,0 +1,279 @@
+     * {@link HttpMethod#knownMethods()}s are set.
+     *
+     * @see #path(String)
+     * @see #pathUnder(String)
+     */
+    public AbstractBindingBuilder methods(HttpMethod... methods) {
+        return methods(ImmutableSet.copyOf(requireNonNull(methods, "methods")));
+    }
+
+    /**
+     * Sets the {@link HttpMethod}s that a {@link Service} will support. If not set,
+     * {@link HttpMethod#knownMethods()}s are set.
+     *
+     * @see #path(String)
+     * @see #pathUnder(String)
+     */
+    public AbstractBindingBuilder methods(Iterable<HttpMethod> methods) {
+        requireNonNull(methods, "methods");
+        checkArgument(!Iterables.isEmpty(methods), "methods can't be empty");
+        this.methods = Sets.immutableEnumSet(methods);
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will consume. If not set, the {@link Service}
+     * will accept all media types.
+     */
+    public AbstractBindingBuilder consumes(MediaType... consumeTypes) {
+        consumes(ImmutableSet.copyOf(requireNonNull(consumeTypes, "consumeTypes")));
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will consume. If not set, the {@link Service}
+     * will accept all media types.
+     */
+    public AbstractBindingBuilder consumes(Iterable<MediaType> consumeTypes) {
+        ensureUniqueMediaTypes(consumeTypes, "consumeTypes");
+        this.consumeTypes = ImmutableSet.copyOf(consumeTypes);
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will produce to be used in
+     * content negotiation. See <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2">Accept header</a>
+     * for more information.
+     */
+    public AbstractBindingBuilder produces(MediaType... produceTypes) {
+        produces(ImmutableSet.copyOf(requireNonNull(produceTypes, "produceTypes")));
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will produce to be used in
+     * content negotiation. See <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2">Accept header</a>
+     * for more information.
+     */
+    public AbstractBindingBuilder produces(Iterable<MediaType> produceTypes) {
+        ensureUniqueMediaTypes(produceTypes, "produceTypes");
+        this.produceTypes = ImmutableSet.copyOf(produceTypes);
+        return this;
+    }
+
+    /**
+     * Returns a newly-created @{@link Route}s  based on the properties of this builder.
+     */
+    final List<Route> buildRouteList() {
+        final ImmutableList.Builder<Route> builder = ImmutableList.builder();
+        pathBuilders.build().forEach(pathBuilder -> addRouteBuilder(pathBuilder, methods));
+
+        routeBuilders.forEach((routeBuilder, methods) -> {
+            builder.add(routeBuilder
+                                .methods(methods)
```

Comment:
I'm a bit confused on the thread about the methods behavior I pointed out - but as @minwoox says, I think the previous code doesn't call this. Can we remove this line so methods aren't overridden for routes added with `get` etc like before?

I'm fine with reexamining the behavior and/or updating javadoc in another PR.

Reply:
Actually, I added this logic intentionally for avoiding break change.
`methods(methods)` are equal to `methods(entry.getValue())`.

https://github.com/line/armeria/blob/352e4b617f67ff96f27a7d32e2194396c0f141b2/core/src/main/java/com/linecorp/armeria/server/AbstractServiceBindingBuilder.java#L446-L448

I totally understand what you are confused on. However, I think we need another discussion to make behavior break change.